RetailCoder.VBE/UI/SourceControl/SourceControlViewViewModel.cs

```diff
@@ -139,6 +169,107 @@ public ISourceControlProvider Provider
+        }
+
+        private void _fileSystemWatcher_Changed(object sender, FileSystemEventArgs e)
+        {
+            // the file system filter doesn't support multiple filters
+            if (!new[] { "cls", "bas", "frm" }.Contains(e.Name.Split('.').Last()))
+            {
+                return;
+            }
+
+            if (!Provider.NotifyExternalFileChanges)
+            {
+                return;
+            }
+
+            Logger.Trace("File system watcher detected file changed");
+            if (_messageBox.Show(RubberduckUI.SourceControl_ExternalModifications, RubberduckUI.SourceControlPanel_Caption,
+                MessageBoxButtons.OKCancel, MessageBoxIcon.Information, MessageBoxDefaultButton.Button1) == DialogResult.OK)
+            {
+                Provider.ReloadComponent(e.Name);
+                UiDispatcher.InvokeAsync(Refresh);
+            }
+        }
+
+        private void _fileSystemWatcher_Renamed(object sender, RenamedEventArgs e)
+        {
+            // the file system filter doesn't support multiple filters
+            if (!new[] { "cls", "bas", "frm" }.Contains(e.Name.Split('.').Last()))
+            {
+                return;
+            }
+
+            if (!Provider.NotifyExternalFileChanges)
+            {
+                return;
+            }
+
+            Logger.Trace("File system watcher detected file renamed");
+            if (_messageBox.Show(RubberduckUI.SourceControl_ExternalModifications, RubberduckUI.SourceControlPanel_Caption,
+                MessageBoxButtons.OKCancel, MessageBoxIcon.Information, MessageBoxDefaultButton.Button1) == DialogResult.OK)
+            {
+                Provider.RemoveFile(e.OldFullPath, true);
+                Provider.AddFile(e.FullPath);
+                UiDispatcher.InvokeAsync(Refresh);
+            }
+        }
+
+        private void _fileSystemWatcher_Deleted(object sender, FileSystemEventArgs e)
+        {
+            // the file system filter doesn't support multiple filters
+            if (!new[] { "cls", "bas", "frm" }.Contains(e.Name.Split('.').Last()))
+            {
+                return;
+            }
+
+            if (!Provider.NotifyExternalFileChanges)
+            {
+                return;
+            }
+
+            Logger.Trace("File system watcher detected file deleted");
+            if (_messageBox.Show(RubberduckUI.SourceControl_ExternalModifications, RubberduckUI.SourceControlPanel_Caption,
+                MessageBoxButtons.OKCancel, MessageBoxIcon.Information, MessageBoxDefaultButton.Button1) == DialogResult.OK)
+            {
+                Provider.RemoveFile(e.FullPath, true);
+                UiDispatcher.InvokeAsync(Refresh);
+            }
+        }
+
+        private void _fileSystemWatcher_Created(object sender, FileSystemEventArgs e)
+        {
+            // the file system filter doesn't support multiple filters
+            if (!new[] { "cls", "bas", "frm" }.Contains(e.Name.Split('.').Last()))
```

Comment:
Perhaps this array should be a `private static readonly IEnumerable<string> VbFileExtensions = new[] { "cls", "bas", "frm" };`, instead of being re-allocated every time a method is called, and copy/pasted in 4 places?